### PR TITLE
Support general marketing consent and subscriptions

### PR DESCRIPTION
**Pull request by dcsaszar**
The naming is taken from the Neoletter UI Profile/Subscriptions tab.

First go. The `fetchMySubscribedTopicIds` helper is duplicated. Ideas for if and where to share code between data connectors are welcome.

Demo content: https://edit.scrivito.com/d0a154d76edf2a7bd991fc658e700a1d~https://dave-neoletter-subscriptions.scrivito-portal-app.pages.dev/subscriptions-027be0e907d46d8e?_scrivito_workspace_id=s204359a88b63a0b&_scrivito_display_mode=view (FYI The icons need an adjustment to allow for boolean values but I consider this out of scope for this PR.)

Neoletter topics: https://edit.neoletter.com/i/d0a154d76edf2a7bd991fc658e700a1d/topics

### CODE REVIEW

**Review by apepper — COMMENTED** [review]
FEEDBACK ([398a115](https://github.com/Scrivito/scrivito-portal-app/pull/289/commits/398a1151a6c180647aa19c234b7a1b6495f2c8fc))

I like the overall idea how you have structured the two endpoints into two other data classes/items.

But we definitely need to work on the UI for this: About about we introduce a `DataFormToggleWidget` that allows to toggle two different states right away (similar to the boolean toggle. This would combine the "data icon", "data form boolean" and "data form button submit" into one.

If you like the idea can you please coordinate with Alex G. to get render-designs and a widget thumbnail for that from him? If it helps, you could submit this `DataFormToggleWidget` as an independent PR. Or if you like you can include it in this PR.

Also it would be good if the content is more explainatory to the visitor. Maybe you can get together with Andreas to improve the wordings of the specific subscriptions page and also add a description for each broad section (e.g. what does a "General marketing consent" consent actually allow).

**Comment by dcsaszar** [comment]
> Also it would be good if the content is more explainatory to the visitor. Maybe you can get together with Andreas to improve the wordings of the specific subscriptions page and also add a description for each broad section (e.g. what does a "General marketing consent" consent actually allow).

FYI I copied the wording 1:1 from
* the examples on https://docs.neoletter.com/topics-259cbad42938db68
* the Neoletter UI

So, if we improve it, we should ideally also improve it in the Neoletter UI and in the docs. CC @aviebke

**Comment by apepper** [comment]
> So, if we improve it, we should ideally also improve it in the Neoletter UI and in the docs.
> 
> image

@dcsaszar Where can I find this UI?

**Comment by dcsaszar** [comment]
> @dcsaszar Where can I find this UI?

https://edit.neoletter.com/i/d0a154d76edf2a7bd991fc658e700a1d/profiles/437d3335797849af99462f0b7a7c1a5c?tab=subscriptions

File: src/Data/Subscription/SubscriptionDataClass.ts
Comment (apepper) [inline src/Data/Subscription/SubscriptionDataClass.ts:30]: If we don't convert `status` to a boolean, but to `status` `"revoked"`/`"given"` instead, we would have not need to implement boolean handling into `DataIconWidget` etc.

Comment (apepper) [reply src/Data/Subscription/SubscriptionDataClass.ts:30]: From the perspective of the end-user there is no big different between unknown and revoked. How about we introduce the artificial status `"given"`/`"notGiven"` to avoid the word "revoke"?

Comment (apepper) [reply src/Data/Subscription/SubscriptionDataClass.ts:30]: Or `"subscribed"`/`"unsubscribed"`.

**Comment by dcsaszar** [comment]
@apepper REVIEW

* The conditional widget from 07623d2 is now unused
* The style is the Bootstrap default and can later be adjusted
* We got some explanatory input for the general marketing consent but haven't found time to update the content yet (CC @aviebke)

image

**Comment by aviebke** [comment]
I was talking to Jan today about how the "General marketing consent" works, explained it in the docs, and I am now waiting for him to digest what I've written and give feedback. Afterwards, we can extract from this text what we think is a useful addition to the Neoletter UI or the portal.

Feel free to take a look at the [working copy](https://edit-beta.scrivito.com/6d226c03f32a8ea4e8ae8e5cbe6c6e2c~https://docs.neoletter.com/profiles-c4fea38cff3bc2be?_scrivito_workspace_id=obd7e08187b64219&_scrivito_display_mode=view#general-marketing-consent) already at this stage.

File: src/Widgets/DataConditionWidget/DataConditionWidgetClass.ts
Comment (apepper) [inline src/Widgets/DataConditionWidget/DataConditionWidgetClass.ts:3]: Please extract (or remove) `DataConditionWidget` from this PR, since it is no longer in use.

File: src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetEditingConfig.ts
Comment (apepper) [inline src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetEditingConfig.ts:28]: Please add:

```ts
  description: 'Default: Checkbox'
```

File: src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetEditingConfig.ts
Comment (apepper) [inline src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetEditingConfig.ts:34]: How about the title `'Immediately submit changed value?'`? Also a description should be added, that this will not apply for new items, only for updating existing items.

File: src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetComponent.tsx
Comment (apepper) [inline src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetComponent.tsx:18]: I suggest to rename `className` to `style`.

File: src/Data/Subscription/SubscriptionDataClass.ts
Comment (apepper) [inline src/Data/Subscription/SubscriptionDataClass.ts:30]: Since we are now showing/updating the value using the `DataFormBooleanWidget` we can switch back to a real boolean here:

```suggestion
          consent: mySubscribedTopicIds.includes(id)
```

This would also make the `dataForNo` and `dataForYes` option in this PR obsolete.

Please also adjust the `consent` flag in the `GeneralMarketingConsent` accordingly.

**Review by apepper — CHANGES_REQUESTED** [review]
FEEDBACK ([61c0b71](https://github.com/Scrivito/scrivito-portal-app/pull/289/commits/61c0b710aa9a76bcf89bf7406ccb82b772f53286))

We're moving in a good direction!

Please see my comments regarding the code.

Content wise please improve the "General marketing consent" section (e.g. with improved wording by Andreas). The goal would be to have a working copy that could be published directly once this PR is approved and merged.

File: src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetComponent.tsx
Comment (krishan) [inline src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetComponent.tsx:21]: Why is `submitOnChange` a property of the `BooleanWidget`?
IMHO it should be a property of the `FormWidget`.

Because having a form where some fields are "submitOnChange" and others are not would be confusing.

Let's keep it simple for the editor and say "a form either submits on change, or i does not".

Or am I missing something?

File: src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetClass.ts
Comment (krishan) [inline src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetClass.ts:9]: Why do we need something complicated like `dataForNo` or `dataForYes`?

Does Neoletter really have such a complex API? Isn't it a simple boolean?

**Comment by krishan** [comment]
General question regarding the approach:

You seem to be working on a intricate upgrade for the boolean toggle, which introduces a lot of complexity into the model (`submitOnChange`, `dataForYes` and such).

Why not just use a simple button ("subscribe" or "unsubscribe")?

File: src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetClass.ts
Comment (krishan) [reply src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetClass.ts:9]: Generally, if something is called ` BooleanWidgetClass`, shouldn't it use booleans?

If the neoletter API needs "special words" instead of booleans, that would be the responsiblity of the data connection.

Comment (apepper) [reply src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetClass.ts:9]: I think this was an intermediate step. Please see https://github.com/Scrivito/scrivito-portal-app/pull/289#discussion_r1492238961 .

File: src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetComponent.tsx
Comment (apepper) [reply src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetComponent.tsx:21]: Good idea: My initial idea was, that only a `DataFormBooleanWidget` without a container could be used. But that would duplicate the save logic to `DataFormBooleanWidget`.

So yeah, let's move `submitOnChange` up to the `DataFormContainerWidget`.

**Comment by apepper** [comment]
> You seem to be working on a intricate upgrade for the boolean toggle, which introduces a lot of complexity into the model (`submitOnChange`, `dataForYes` and such).
> 
> Why not just use a simple button ("subscribe" or "unsubscribe")?

A `subscribe`/`unsubscribe` button was Daves initial implementation (see https://github.com/Scrivito/scrivito-portal-app/pull/289#issuecomment-1936141494). But it relied on a `DataConditionWidget` that original also relied on a string representation of a boolean value. One could change this boolean value to a string like `'given'`/`'notGiven'` to avoid the "string representation of boolean value".

But to me a boolean-toggle seems to be more "state of the art" for subscriptions, that's why I ask Dave to go in this direction instead.

**Comment by krishan** [comment]
> it relied on a DataConditionWidget that original also relied on a string representation of a boolean value

Side Note: As you know, my plan is to allow for DataLocators to refer to attributes (see: https://github.com/infopark/scrivito_js/issues/10255 ). AFAICS this would also help us here: If the `DataConditionWidget` finds that its datalocator points to an attribute of type boolean, it just uses that. 

So that would have solve this problem here, right?

**Comment by apepper** [comment]
> Side Note: As you know, my plan is to allow for DataLocators to refer to attributes (see: [infopark/scrivito_js#10255](https://github.com/infopark/scrivito_js/issues/10255) ). AFAICS this would also help us here: If the `DataConditionWidget` finds that its datalocator points to an attribute of type boolean, it just uses that.
> 
> So that would have solve this problem here, right?

With the implementation of `DataConditionWidget` as it was proposed by Dave this would solve the problem here. But I'm not yet 100% sure, if the proposed `DataConditionWidget` would work well for any other non-boolean attribute.

File: src/Data/GeneralMarketingConsent/GeneralMarketingConsentEditingConfig.ts
Comment (apepper) [inline src/Data/GeneralMarketingConsent/GeneralMarketingConsentEditingConfig.ts:7]:
```suggestion
    isConsentGiven: { title: 'Is consent given?' },
```

File: src/Data/Subscription/SubscriptionEditingConfig.ts
Comment (apepper) [inline src/Data/Subscription/SubscriptionEditingConfig.ts:9]:
```suggestion
    isConsentGiven: { title: 'Is consent given?' },
```

File: src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetComponent.tsx
Comment (apepper) [inline src/Widgets/DataFormBooleanWidget/DataFormBooleanWidgetComponent.tsx:24]: How about this?

```ts
const classNames = ['mb-3', 'form-check']
if (widget.get('style') === 'switch') { classNames.push('form-switch') }

// ...
className={classNames.join(' ')}
```

**Review by apepper — CHANGES_REQUESTED** [review]
FEEDBACK ([c0ce476](https://github.com/Scrivito/scrivito-portal-app/pull/289/commits/c0ce476a57c4b8ba586dbf17d7a4edf61d8d30ec))

Nice 🚀!

Almost there. Only some minor details can be improved (see comments).

**Comment by dcsaszar** [comment]
REVIEW

The content for "MAILINGS NOT LINKED TO A NEWSLETTER TOPIC" has been approved by Andreas.

**Review by apepper — APPROVED** [review]
LGTM ([a72058f](https://github.com/Scrivito/scrivito-portal-app/pull/289/commits/a72058facef15c4a07f52d1347f4aebceb7699e2))

Very nice 🚀!

**Pull request merged**
